File: native/ios/Circuul.swift
import Foundation

/// Circuul iOS client — Universal Links first, clipboard last.
/// Call `Circuul.shared.match(...)` on first open; never block launch.
public final class Circuul {
    public static let shared = Circuul()

    private var appToken: String = ""
    private var apiBase: String = ""
    private let defaults = UserDefaults.standard
    private let installKey = "circuul_install_id"
    private let matchedKey = "circuul_matched"

    public func configure(appToken: String, apiBase: String) {
        self.appToken = appToken
        self.apiBase = apiBase.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
    }

    public func installId() -> String {
        if let existing = defaults.string(forKey: installKey) { return existing }
        let id = UUID().uuidString
        defaults.set(id, forKey: installKey)
        return id
    }

    /// Fire-and-forget safe. Prefer passing Universal Link `code`; clipboard is last resort.
    public func match(
        platform: String = "ios",
        code: String? = nil,
        clipboardCode: String? = nil,
        idfv: String? = nil,
        completion: (([String: Any]) -> Void)? = nil
    ) {
        if defaults.bool(forKey: matchedKey) {
            completion?(["attributed": false, "reason": "already_matched"])
            return
        }
        guard let url = URL(string: "\(apiBase)/circuul/match") else {
            completion?(["attributed": false, "reason": "bad_url"])
            return
        }
        var body: [String: Any] = [
            "app_token": appToken,
            "platform": platform,
            "install_id": installId(),
        ]
        if let code = code { body["code"] = code }
        if let clipboardCode = clipboardCode { body["clipboard_code"] = clipboardCode }
        if let idfv = idfv { body["idfv"] = idfv }

        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = try? JSONSerialization.data(withJSONObject: body)

        URLSession.shared.dataTask(with: req) { data, _, _ in
            self.defaults.set(true, forKey: self.matchedKey)
            guard let data = data,
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let payload = json["data"] as? [String: Any] else {
                completion?(["attributed": false, "reason": "network_error"])
                return
            }
            completion?(payload)
        }.resume()
    }
}

File: native/android/Circuul.kt
package com.thspian.circuul

import android.content.Context
import android.os.Handler
import android.os.Looper
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL
import java.util.UUID
import kotlin.concurrent.thread

/**
 * Circuul Android client — Play Install Referrer first, clipboard last.
 * Call [match] on first open; never block the main thread / launch.
 */
object Circuul {
    private const val PREFS = "circuul"
    private const val INSTALL_KEY = "install_id"
    private const val MATCHED_KEY = "matched"

    private var appToken: String = ""
    private var apiBase: String = ""

    fun configure(appToken: String, apiBase: String) {
        this.appToken = appToken
        this.apiBase = apiBase.trimEnd('/')
    }

    fun installId(context: Context): String {
        val prefs = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
        val existing = prefs.getString(INSTALL_KEY, null)
        if (existing != null) return existing
        val id = UUID.randomUUID().toString()
        prefs.edit().putString(INSTALL_KEY, id).apply()
        return id
    }

    fun match(
        context: Context,
        platform: String = "android",
        androidReferrer: String? = null,
        code: String? = null,
        clipboardCode: String? = null,
        gaid: String? = null,
        onResult: ((JSONObject) -> Unit)? = null,
    ) {
        val prefs = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
        if (prefs.getBoolean(MATCHED_KEY, false)) {
            onResult?.invoke(JSONObject(mapOf("attributed" to false, "reason" to "already_matched")))
            return
        }

        thread {
            val result = try {
                val body = JSONObject()
                    .put("app_token", appToken)
                    .put("platform", platform)
                    .put("install_id", installId(context))
                if (androidReferrer != null) body.put("android_referrer", androidReferrer)
                if (code != null) body.put("code", code)
                if (clipboardCode != null) body.put("clipboard_code", clipboardCode)
                if (gaid != null) body.put("gaid", gaid)

                val conn = (URL("$apiBase/circuul/match").openConnection() as HttpURLConnection)
                conn.requestMethod = "POST"
                conn.setRequestProperty("Content-Type", "application/json")
                conn.doOutput = true
                conn.outputStream.use { it.write(body.toString().toByteArray()) }
                val text = conn.inputStream.bufferedReader().readText()
                val json = JSONObject(text)
                json.optJSONObject("data")
                    ?: JSONObject(mapOf("attributed" to false, "reason" to "invalid_response"))
            } catch (_: Exception) {
                JSONObject(mapOf("attributed" to false, "reason" to "network_error"))
            }

            prefs.edit().putBoolean(MATCHED_KEY, true).apply()
            Handler(Looper.getMainLooper()).post { onResult?.invoke(result) }
        }
    }
}
